### PR TITLE
Use percentage to represent memory/cpu usage

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Review by kezhenxu94 — DISMISSED** [review]

**Review by davidzollo — DISMISSED** [review]
+1

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/OSUtils.java [hunk 1]
```diff
@@ -466,23 +433,25 @@ public static int getProcessID() {
...
+                    maxCpuLoadAvgThreshold);
             return true;
         }
 
-        if (availablePhysicalMemorySize < reservedMemory) {
+        if (freeMemoryPercentage < reservedMemoryThreshold) {
             log.warn(
-                    "Current available memory {}G is too low, reserved.memory={}G", maxCpuLoadAvg, reservedMemory);
+                    "Current available memory percentage{} is too low, reserved.memory={}", freeMemoryPercentage,
+                    reservedMemoryThreshold);
```
Comment (Radeity) [inline dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/OSUtils.java:454]: The unit of `reservedMemory ` is still GB, if you also want to change the memory representation to percentage, you may have to modify the check in `getServerStatus `.

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/task/WorkerHeartBeatTask.java [hunk 2]
```diff
@@ -90,16 +88,16 @@ public void writeHeartBeat(WorkerHeartBeat workerHeartBeat) {
...
-    public int getServerStatus(double loadAverage,
-                               double maxCpuloadAvg,
+    public int getServerStatus(double cpuUsagePercentage,
+                               double maxCpuUsePercentage,
                                double availablePhysicalMemorySize,
                                double reservedMemory,
                                int workerExecThreadCount,
                                int workerWaitingTaskCount) {
-        if (loadAverage > maxCpuloadAvg || availablePhysicalMemorySize < reservedMemory) {
+        if (cpuUsagePercentage > maxCpuUsePercentage || availablePhysicalMemorySize < reservedMemory) {
```
Comment (Radeity) [inline dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/task/WorkerHeartBeatTask.java]: Here, it still compares the physical memory with reserved memory, their unit are both GB.

File: dolphinscheduler-worker/src/main/resources/application.yaml [hunk 3]
```diff
@@ -48,9 +48,9 @@ worker:
   tenant-auto-create: true
   #Scenes to be used for distributed users.For example,users created by FreeIpa are stored in LDAP.This parameter only applies to Linux, When this parameter is true, worker.tenant.auto.create has no effect and will not automatically create tenants.
   tenant-distributed-user: false
-  # worker max cpuload avg, only higher than the system cpu load average, worker server can be dispatched tasks. default value -1: the number of cpu cores * 2
-  max-cpu-load-avg: -1
-  # worker reserved memory, only lower than system available memory, worker server can be dispatched tasks. default value 0.3, the unit is G
+  # worker max cpuload avg, only higher than the system cpu load average, worker server can be dispatched tasks. default value 1: will use 100% cpu.
+  max-cpu-load-avg: 1
+  # worker reserved memory, only lower than system available memory, worker server can be dispatched tasks. default value 0.3, only the available memory is higher than 30%, worker server can receive task.
   reserved-memory: 0.3
```
Comment (Radeity) [inline dolphinscheduler-worker/src/main/resources/application.yaml:54]: Maybe you can modify the doc of `reserved-memory`.

**Review by Radeity — CHANGES_REQUESTED** [review]
Hi, @ruanwenjun , i just wonder why we use percentage representation? For me, physical memory is better.

File: dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/OSUtils.java (hunk 1 above)
Comment (ruanwenjun) [reply dolphinscheduler-common/src/main/java/org/apache/dolphinscheduler/common/utils/OSUtils.java:454]: Thanks, done.

File: dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/task/WorkerHeartBeatTask.java (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-worker/src/main/java/org/apache/dolphinscheduler/server/worker/task/WorkerHeartBeatTask.java]: Done

File: dolphinscheduler-worker/src/main/resources/application.yaml (hunk 3 above)
Comment (ruanwenjun) [reply dolphinscheduler-worker/src/main/resources/application.yaml:54]: Thanks, done.

**Comment by ruanwenjun** [comment]
> Hi, @ruanwenjun , i just wonder why we use percentage representation? For me, physical memory is better.

Since the instance's memory/cpu might different, use percentage is more suitable, otherwise we need to set different config.

**Comment by Radeity** [comment]
> Since the instance's memory/cpu might different, use percentage is more suitable, otherwise we need to set different config.

If physical memory of different worker vary significantly, such as workerA has 8G memory and workerB has 64G, it's better to set different config, WDYT? 

Anyway, we can still use different percentage to represent it if you think percentage is more intuitive :D

**Comment by ruanwenjun** [comment]
> > Since the instance's memory/cpu might different, use percentage is more suitable, otherwise we need to set different config.
> 
> If physical memory of different worker vary significantly, such as workerA has 8G memory and workerB has 64G, it's better to set different config, WDYT?
> 
> Anyway, we can still use different percentage to represent it if you think percentage is more intuitive :D

I think we may more concern about the percentage of usage.

**Comment by EricGao888** [comment]
> > Since the instance's memory/cpu might different, use percentage is more suitable, otherwise we need to set different config.
> 
> If physical memory of different worker vary significantly, such as workerA has 8G memory and workerB has 64G, it's better to set different config, WDYT?
> 
> Anyway, we can still use different percentage to represent it if you think percentage is more intuitive :D

Actually I think both make sense to me. May I ask whether it is possible to keep both of them and give users an option to configure whether by physical memory or by percentage?

**Comment by ruanwenjun** [comment]
> > > Since the instance's memory/cpu might different, use percentage is more suitable, otherwise we need to set different config.
> > 
> > 
> > If physical memory of different worker vary significantly, such as workerA has 8G memory and workerB has 64G, it's better to set different config, WDYT?
> > Anyway, we can still use different percentage to represent it if you think percentage is more intuitive :D
> 
> Actually I think both make sense to me. May I ask whether it is possible to keep both of them and give users an option to configure whether by physical memory or by percentage?

This is possible, we can provide different strategy, e.g percentage, absolute value. This will need to change the whole config.
```yml
avoid-overload:
    strategy: Percentage
    max-memory-usage: 70%
    max-cpu-usage: 80%    
    # strategy: AbsoluteValue 
    # max-memory-usage: 6G
    # max-cpu-usage: 6C
```

If this is needed, I will improve this in the next PR.

**Comment by Radeity** [comment]
> 
> This is possible, we can provide different strategy, e.g percentage, absolute value. This will need to change the whole config.
> 
> ```yaml
> avoid-overload:
>     strategy: Percentage
>     max-memory-usage: 70%
>     max-cpu-usage: 80%    
>     # strategy: AbsoluteValue 
>     # max-memory-usage: 6G
>     # max-cpu-usage: 6C
> ```
> 
> If this is needed, I will improve this in the next PR.

Maybe we can simply provide two settings, if satisfy both, we consider the node is overload, like the memory usage is higher than 70% and reserved memory is less than 6G.

**Comment by EricGao888** [comment]
> > > > Since the instance's memory/cpu might different, use percentage is more suitable, otherwise we need to set different config.
> > > 
> > > 
> > > If physical memory of different worker vary significantly, such as workerA has 8G memory and workerB has 64G, it's better to set different config, WDYT?
> > > Anyway, we can still use different percentage to represent it if you think percentage is more intuitive :D
> > 
> > 
> > Actually I think both make sense to me. May I ask whether it is possible to keep both of them and give users an option to configure whether by physical memory or by percentage?
> 
> This is possible, we can provide different strategy, e.g percentage, absolute value. This will need to change the whole config.
> 
> ```yaml
> avoid-overload:
>     strategy: Percentage
>     max-memory-usage: 70%
>     max-cpu-usage: 80%    
>     # strategy: AbsoluteValue 
>     # max-memory-usage: 6G
>     # max-cpu-usage: 6C
> ```
> 
> If this is needed, I will improve this in the next PR.

+1

**Review by caishunfeng — APPROVED** [review]
:+1:

**Pull request merged**